src/NodeInterface.php

```diff
@@ -0,0 +1,29 @@
+<?php
+
+namespace Dhii\Data\Tree;
+
+use Dhii\Data\ParentAwareInterface;
+use Dhii\Data\ChildrenAwareInterface;
+use Dhii\Data\ValueAwareInterface;
+use Dhii\Data\KeyAwareInterface;
+
+/**
+ * Something that can be a tree node.
+ *
+ * @since [*next-version*]
+ */
+interface NodeInterface extends
```

Comment:
I suggest breaking down this interface into more granular interfaces. Not every node requires a reference to the parent or a "key" value. In some instances, children are also not required and `getChildren()` returning an empty list is not always conceptually correct.

I recommend having a bottom-most interface that is just value aware, such as a `ElementInterface`. An element in a tree is at most required to have a value, since a tree's purpose is first and foremost a data structure, hence implying data storage.

This interface can then be extended as follows:

* `NodeInterface` would be what defines an actual "node" as understood in compute science: an object that holds a value and has references to child elements.
* `ParentAwareElementInterface` is an element that is aware of a parent, i.e. allows the retrieval of a reference to another `NodeInterface` (not `ElementInterface` in my opinion, since if an element is a parent, conceptually it means that it has children).

A `ParentAwareNodeInterface` can also be included, that extends `NodeInterface` and `ParentAwareElementInterface`, thus representing an element with both parent and child references.

----

In my opinion, key awareness should not be a concern for this package.

**Scenario 1**

Elements can have objects as values, which themselves could be key aware.

```
[ElementInterface]->getValue()->getKey();
```

**Scenario 2**

Consumers that depend on an element that is key aware should keep in mind that a key is conceptually a node of its own. In mapped structures, it can imply only a single child node (the mapped value), but in a tree hierarchy, this restriction is not imposed.

A good example is a DOM-like structure, where attributes are also nodes. A DOM node can have child nodes for each attribute. These child nodes would hold the attribute key as the element value.

```
$attrNode->getValue(); // returns the attribute key
```

Each of these attribute nodes would then have their own children whose value holds the actual attribute value. This also allows for the same attribute key to have multiple values.

```
[NodeInterface]
├── [NodeInterface] { value: "my-attr" }
│   └── [ParentAwareElementInterface] {value: "my-value"}
└── [NodeInterface] { value: "another-attr" }
    ├── [ParentAwareElementInterface] {value: "first-value"}
    └── [ParentAwareElementInterface] {value: "second-value"}
```